contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -885,6 +917,243 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
+				break;
+			}
+		}
+		if (!type_match)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+				 		 format_type_be(inputTypId), i+1, special_func->funcname)));
+		}
+	}
+
+	return true;
+}
+
+/*
+ * tsql_func_select_candidate_for_special_func()
+ *
+ * For functions present in special function list, and try to find best candidate 
+ * based on matching return type. Also throw error in case of invalid argument data type.
+ */
+static FuncCandidateList
+tsql_func_select_candidate_for_special_func(List *names, int nargs, Oid *input_typeids, FuncCandidateList candidates)
+{
+	FuncCandidateList			current_candidate, best_candidate;
+	Oid 						expr_result_type;
+	char					   *proc_nsname;
+	char					   *proc_name;
+	bool						is_func_validated;
+	int							ncandidates;
+	Relation					relsetting;
+	HeapTuple					tuple;
+	ScanKeyData					scankey;
+	SysScanDesc					scan;
+	bool						isnull;
+	Oid							retType;
+	Datum						datum;
+
+	DeconstructQualifiedName(names, &proc_nsname, &proc_name);
+
+	is_func_validated = validate_special_function(proc_nsname, proc_name, nargs, input_typeids);
+
+	/* Return NULL if function is not a special function */
+	if (!is_func_validated)
+		return NULL;
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* function based logic to decide return type */
+	expr_result_type = InvalidOid;
+	if (strlen(proc_name) == 4 && strcmp(proc_name,"trim") == 0)
+	{
+		if ((*common_utility_plugin_ptr->is_tsql_nvarchar_datatype)(input_typeids[0])
+			|| (*common_utility_plugin_ptr->is_tsql_nchar_datatype)(input_typeids[0]))
+		{
+			expr_result_type = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid) ("nvarchar");	
+		}
+		else if ((*common_utility_plugin_ptr->is_tsql_varchar_datatype)(input_typeids[0])
+				|| (*common_utility_plugin_ptr->is_tsql_bpchar_datatype)(input_typeids[0])
+				|| input_typeids[0] == UNKNOWNOID)
+		{
+			expr_result_type = get_sys_varcharoid();
+		}
+	}
+
+	if (!OidIsValid(expr_result_type))
+		return NULL;
+
+	/* Get the candidate with matching return type */
+	ncandidates = 0;
+	best_candidate = NULL;
+	relsetting = table_open(ProcedureRelationId, AccessShareLock);
+	for (current_candidate = candidates;
+			current_candidate != NULL;
+			current_candidate = current_candidate->next)
+	{
+		ScanKeyInit(&scankey,
+					Anum_pg_proc_oid,
+					BTEqualStrategyNumber, F_OIDEQ,
+					ObjectIdGetDatum(current_candidate->oid));
+	
+		scan = systable_beginscan(relsetting, ProcedureOidIndexId, true, NULL, 1, &scankey);
+		tuple = systable_getnext(scan);
+		if (HeapTupleIsValid(tuple))
+		{
+			datum = heap_getattr(tuple, Anum_pg_proc_prorettype, RelationGetDescr(relsetting), &isnull);
+			if (!isnull)
+			{
+				retType = DatumGetObjectId(datum);	
```

Comment:
Can't we directly use `get_func_rettype`.

Reply:
+1

Reply:
Done